File: libs/login-api-2/src/main/java/com/instructure/loginapi/login/api/MobileVerifyAPI.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */
package com.instructure.loginapi.login.api;

import com.instructure.canvasapi2.StatusCallback;
import com.instructure.canvasapi2.utils.APIHelper;
import com.instructure.canvasapi2.utils.ApiPrefs;
import com.instructure.canvasapi2.utils.Logger;
import com.instructure.loginapi.login.model.DomainVerificationResult;
import com.instructure.canvasapi2.utils.CanvasAuthenticator;

import java.io.IOException;

import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Query;

public class MobileVerifyAPI {

    private static Retrofit getAuthenticationRetrofit() {

        final String userAgent = ApiPrefs.getUserAgent();

        OkHttpClient httpClient = new OkHttpClient.Builder()
                .addInterceptor(new Interceptor() {
                    @Override
                    public Response intercept(Chain chain) throws IOException {
                        if (!userAgent.equals("")) {
                            Request request = chain.request().newBuilder()
                                    .header("User-Agent", userAgent)
                                    .cacheControl(CacheControl.FORCE_NETWORK)
                                    .build();
                            return chain.proceed(request);
                        } else {
                            return chain.proceed(chain.request());
                        }
                    }
                })
                .build();

        return new Retrofit.Builder()
                .baseUrl("https://canvas.instructure.com/api/v1/")
                .client(httpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
    }

    interface OAuthInterface {
        @GET("mobile_verify.json")
        Call<DomainVerificationResult> mobileVerify (@Query(value = "domain", encoded = false) String domain, @Query("user_agent") String userAgent);
    }

    public static void mobileVerify(String domain, StatusCallback<DomainVerificationResult> callback) {
        if (APIHelper.INSTANCE.paramIsNull(callback, domain)) {
            return;
        }

        final String userAgent = ApiPrefs.getUserAgent();
        if (userAgent.equals("")) {
            Logger.d("User agent must be set for this API to work correctly!");
            return;
        }

        OAuthInterface oAuthInterface = getAuthenticationRetrofit().create(OAuthInterface.class);
        oAuthInterface.mobileVerify(domain, userAgent).enqueue(callback);
    }
}

File: libs/login-api-2/src/main/java/com/instructure/loginapi/login/api/MobileVerifyAPI.kt
/*
 * Copyright (C) 2020 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.instructure.loginapi.login.api

import com.instructure.canvasapi2.StatusCallback
import com.instructure.canvasapi2.utils.APIHelper.paramIsNull
import com.instructure.canvasapi2.utils.ApiPrefs
import com.instructure.canvasapi2.utils.Logger.d
import com.instructure.canvasapi2.utils.RemoteConfigParam
import com.instructure.canvasapi2.utils.RemoteConfigUtils
import com.instructure.loginapi.login.model.DomainVerificationResult
import com.instructure.loginapi.login.util.BaseConfigurations
import okhttp3.CacheControl
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.GET
import retrofit2.http.Query

object MobileVerifyAPI {

    internal interface OAuthInterface {
        @GET("mobile_verify.json")
        fun mobileVerify(
            @Query(value = "domain", encoded = false) domain: String?,
            @Query("user_agent") userAgent: String?
        ): Call<DomainVerificationResult>
    }

    private fun getAuthenticationRetrofit(domain: String?): Retrofit {
        val httpClient = OkHttpClient.Builder()
            .addInterceptor { chain ->
                val request = chain.request().newBuilder()
                    .header("User-Agent", ApiPrefs.userAgent)
                    .cacheControl(CacheControl.FORCE_NETWORK)
                    .build()
                chain.proceed(request)
            }.build()

        val mobileVerifyBetaEnabled = RemoteConfigUtils.getString(
            RemoteConfigParam.MOBILE_VERIFY_BETA_ENABLED
        )?.equals("true", ignoreCase = true)
            ?: false

        // We only want to switch over to the beta mobile verify domain if the remote firebase config is true
//        val baseUrl = if (mobileVerifyBetaEnabled && domain?.contains(".beta.") == true) {
//            "https://canvas.beta.instructure.com/api/v1/"
//        } else {
//            "${BaseConfigurations.baseUrl}/api/v1/"
////                "https://canvas-test.emeritus.org/api/v1/"
//        }

        val baseUrl = "${BaseConfigurations.baseUrl}/api/v1/"

        return Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(httpClient)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }

    fun mobileVerify(domain: String?, callback: StatusCallback<DomainVerificationResult>) {
        if (paramIsNull(callback, domain)) {
            return
        }
        if (ApiPrefs.userAgent == "") {
            d("User agent must be set for this API to work correctly!")
            return
        }
        val oAuthInterface = getAuthenticationRetrofit(domain).create(OAuthInterface::class.java)
        oAuthInterface.mobileVerify(domain, ApiPrefs.userAgent).enqueue(callback)
    }
}
